testnet/stacks-node/src/run_loop/neon.rs

```diff
@@ -313,7 +317,30 @@ impl RunLoop {
         // prepare to fetch the first reward cycle!
         target_burnchain_block_height = burnchain_height + pox_constants.reward_cycle_length as u64;
 
+        let should_keep_running = Arc::new(AtomicBool::new(true));
+        let keep_running_writer = should_keep_running.clone();
+
+        termination::set_handler(move || {
+            keep_running_writer.store(false, Ordering::SeqCst);
+        })
+        .expect("Error setting termination handler");
+
         loop {
+            // Orchestrating graceful termination
+            if !should_keep_running.load(Ordering::SeqCst) {
+                info!("Terminating process");
+                // Terminate chains coordinator
+                info!("Stopping chains-coordinator");
+                coordinator_senders.stop_chains_coordinator();
+                // Terminate relayer
+                info!("Stopping relayer");
+                node.terminate_relayer();
+                // Commands handling is async, we'll sleep to make sure
+                // that the exit commands sent are ingested.
+                sleep(Duration::from_secs(10));
+                info!("Stopping stacks-node");
+                break;
+            }
```

Comment:
Is there anything else that has to happen here?  Assuming all threads gracefully exit, are the node's databases all left in a consistent state such that the node can be started up again?

Reply:
I think the only thread that won't _necessarily_ exit gracefully in the above implementation is the `p2p` thread and the `dns-resolver`. The `p2p` thread should exit gracefully once the `relayer` thread has exited (hanging up the relayer channel should cause the `p2p` thread to gracefully exit), but it may take it some time until the network receives data that needs to be passed to the relayer (which may not happen at all once the coordinator stops and the relayer stops). So this probably means that we need to submit an explicit exit command to the p2p thread somehow (the p2p thread _could_ leave bad db state if exited improperly).

Reply:
Agreed.  The `dns-resolver` doesn't use any DB state, but the `p2p` thread can store blocks, microblocks, and transactions sent to it via the RPC interface.  So, it will need to be explicitly told to shut down.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2433/commits/b2cc9e3b1fb15402af9191467a5197a3a160c69d.
Also, instead of terminating the relayer from the main thread, the p2p thread is now in charge of sending the Exit command to the relayer, once all the network results from the ongoing p2p runloop epoch have been processed